locator/network_topology_strategy.cc

```diff
@@ -336,39 +335,128 @@ future<tablet_map> network_topology_strategy::allocate_tablets_for_new_table(sch
+                continue;
+            }
 
-        tablet_replica_set replicas;
-        for (auto&& ep : tracker.replicas()) {
-            replicas.emplace_back(tablet_replica{ep, load.next_shard(ep)});
+            if (dc_rf > dr_load.node_count) {
+                auto cmp = [] (const rack_load& a, const rack_load& b) {
+                    return a.load > b.load;
+                };
+
+                // Track new racks, with no replicas on them, in random order
+                const auto& all_dc_racks = dc_rack_nodes.at(dc);
+                auto rack_names = boost::copy_range<std::vector<sstring>>(all_dc_racks | boost::adaptors::map_keys);
+                std::shuffle(rack_names.begin(), rack_names.end(), rnd_engine);
+                std::make_heap(dr_load.racks.begin(), dr_load.racks.end(), cmp);
+                for (const auto& rack : rack_names) {
+                    if (!replicas_per_dc[dc].contains(rack)) {
+                        dr_load.racks.emplace_back(rack, 0);
+                        std::push_heap(dr_load.racks.begin(), dr_load.racks.end(), cmp);
+                    }
+                }
+                // Track all nodes with no replicas on them, per rack.
+                std::unordered_map<sstring, std::vector<locator::host_id>> candidate_nodes_per_rack;
+                for (const auto& rack_name : rack_names) {
+                    const auto& existing = replicas_per_dc[dc][rack_name];
+                    auto& rack = candidate_nodes_per_rack[rack_name];
+                    for (const auto& node : all_dc_racks.at(rack_name)) {
+                        const auto& host_id = node->host_id();
+                        if (!existing.contains(host_id)) {
+                            rack.emplace_back(host_id);
+                        }
+                    }
+                    std::shuffle(rack.begin(), rack.end(), rnd_engine);
+                }
+                size_t remaining = dc_rf - dr_load.node_count;
+                if (cur_tablets) {
+                    tablet_logger.debug("Allocating additional tablet replicas in dc={} allocated={} rf={}", dc, dr_load.node_count, dc_rf);
+                }
+                while (remaining) {
+                    co_await coroutine::maybe_yield();
+                    if (dr_load.racks.empty()) {
+                        on_internal_error(tablet_logger, format("Cannot allocate table replica in dc={}: no available racks", dc));
+                    }
+                    std::pop_heap(dr_load.racks.begin(), dr_load.racks.end(), cmp);
+                    auto& rl = dr_load.racks.back();
+                    auto& rack = candidate_nodes_per_rack[rl.rack_name];
+                    if (rack.empty()) {
+                        dr_load.racks.pop_back();
+                        continue;
+                    }
+                    const auto& host_id = rack.back();
+                    replicas.emplace_back(tablet_replica{host_id, load.next_shard(host_id)});
```

Comment:
We need to keep racks in the same order in the replica list. Migrations only happen in the same ordinal position within the replica list, and this keeps  the migration within the rack.

So we need to special-case dc_rf == all_dc_racks.size(), and in that special case, have a 1:1 relationship between the ordinal position in replicas and the ordinal position in all_dc_racks.

Reply:
We also need to document
 - when adding a datacenter, add it all at once
 - prefer having rack replication factor equal to the number of racks

Reply:
done in v2 uisng ordinal rack order.
nodes within a rack are consumed so to fill least loaded nodes first, yet nodes with equal load are shuffled.
added patch to use a tablet_id % number of racks in dc for selecting the primary replica for repair (as the first replica in the rack - hence the shuffling)